Added imread, imsave and synchronize functionality

thawn opened the pull request:
You can now just use imread(filename) and the library uses a local file on the cluster if it finds a matching one, or it downloads the file to a temporary location from the fileserver and accesses that one.

Similarly, imsave(filename, data), saves the file if the path is accessible. Otherwise, it saves to a temp file and then copies that to the final location (on the fileserver or project space) using the datamover tools.

The best way to test is to install the test singularity image for biapol-taurus:
```bash
git clone https://gitlab.mn.tu-dresden.de/bia-pol/singularity-devbio-napari.git
cd singularity-devbio-napari
./install.sh biapol_taurus
```
and then open a notebook with the `devbio-napari_biapol_taurus` kernel.

inside the notebook run:
```python
from biapol_taurus import ProjectFileTransfer
from skimage.io import imshow

# Define the source directory. /grp/g_biapol/ is the mount point at the export node. HPCsupport can tell us this:
source_dir = "/grp/g_biapol/"

# Define the target directory. /projects/p_bioimage/ is the project space. HPCsupport told us that when we applied for the project:
target_dir = "/projects/p_bioimage/"

# We setup a connection:
pft = ProjectFileTransfer(source_dir, target_dir)

image = pft.imread('data/test/blobs19.tif')
# show image
imshow(image)
```
```python
pft.imsave('/projects/p_bioimage/till/blobs_saved.tif', image)
```

thawn committed:
switched to using taurus-datamover

thawn committed:
switch to using pathlib where possible

thawn committed:
removed obsolete import

thawn committed:
added list_fileserver_files

thawn committed:
removed unused import

thawn committed:
added imread and imsave

thawn committed:
clean up imports

thawn committed:
fixed dependencies

thawn committed:
fixed bug in remove_file

thawn committed:
preserve relative file structure

thawn committed:
fixed path for list_fileserver_files

thawn committed:
return target file

thawn committed:
make sure we get the correct file

thawn committed:
make sure filename is a string in remove_file

enables passing Pathh objects to remove_file

thawn committed:
fixed target path for get_file

thawn committed:
fixed target directory problem

thawn committed:
fixed warning layout

thawn committed:
fixed warning layout

thawn committed:
added timeout to list_fileserver

thawn committed:
no timeout for remove_file

thawn committed:
fixed target directory problem

thawn committed:
improved documentation and method order

thawn committed:
more semantic name

thawn committed:
fixed cache bug

thawn committed:
clearer names

thawn committed:
bumped version

thawn committed:
switch to taurus_datamover.CacheWorkspace

thawn committed:
bumped version

thawn committed:
fixed datatype

thawn committed:
fix temp dir for imsave

thawn committed:
fixed imwrite

thawn committed:
more sensible default behavior for imsave

thawn committed:
more verbose output

thawn committed:
switched to using save_to_project in imsave

thawn committed:
fixed imsave

haesleinhuepf commented on setup.py:
```diff
@@ -5,16 +5,15 @@
+    url="https://github.com/BiAPoL/biapol-taurus",
+    packages=setuptools.find_packages(exclude=["docs"]),
     include_package_data=True,
-    install_requires=["numpy!=1.19.4", "stackview"],
```
don't we need requirements? How about specifying `taurus_datamover` ?

thawn replied:
added taurus-datamover and scikit-image as dependencies

haesleinhuepf commented on biapol_taurus/_project_file_transfer.py:
```diff
@@ -129,58 +276,27 @@ def remove_file(self, filename, timeout_in_s: float = 20, wait_for_finish: bool
+            filename = Path(filename)
 
-        self._run_command([self.dtrm, filename])
+        proc = self.dm.dtrm('-r', str(filename))
```
I'm not a big fan of such code. I cannot speak out `proc dm dtrm`. Please rename variable names to english words if possible.

thawn replied:
😳 

I fixed `dm` and `proc`. However, I cannot fix `dtrm` because that is the filename of the executable. taurus_datamover gets these attributes automatically from the filenames (see the code there), which has the advantage that it automatically extends its capabilities when new commands are added by the ZIH (or by us in the containers).

haesleinhuepf commented on biapol_taurus/_project_file_transfer.py:
```diff
@@ -1,98 +1,233 @@
 
-    def __init__(self, source_mount: str, target_project_space: str):
+    def __init__(self, source_fileserver_dir: str, target_project_space_dir: str,
+                 dm_path: str = '/sw/taurus/tools/slurmtools/default/bin/'):
```
please rename parameter names to english words "dm_path" is also not extremely self-explanatory.

haesleinhuepf commented on biapol_taurus/_project_file_transfer.py:
```diff
@@ -1,98 +1,233 @@
+from taurus_datamover import Datamover, CacheWorkspace, waitfor, save_to_project
+
+
+class DangerousOperationException(IOError):
+    '''Exception that is raised when a dangerous operatino is called without im_sure=True
+    '''
+    pass
```
Do we really need this? It looks very dangerous

thawn replied:
yes, especially the operatiNO ;-)

Some operations (i.e. rsync --delete) could cause data loss or (like syncing the entire fileserver with the project space even without --delete) mess up the data structure. I think such operations require an extra confirmation.

I am open for suggestions on how to better implement this :)

A cosmetic change would be to rename it to `ConfirmationRequiredException`.

thawn replied:
I now implemented the cosmetic change. Does that look o.k. to you?

haesleinhuepf commented on biapol_taurus/_project_file_transfer.py:
```diff
@@ -1,98 +1,233 @@
+                target_path = str(self.source_fileserver_dir / filename)
+        return save_to_project(imsave, str(target_path), data, *args, **kw)
+
+    def sync_with_fileserver(self, direction: str = 'from fileserver', delete: bool = False,
```
wow, this is super cool!

haesleinhuepf commented on biapol_taurus/_project_file_transfer.py:
```diff
@@ -1,98 +1,233 @@
+                 wait_for_finish: bool = True) -> Path:
+        '''Ensures that the computing node has access to a file. If necessary, the file is retrieved from a mounted fileserver share.
+
+        Before transferring the file, local directories are checked in the following order: 1. filename (in case the user gave a path to an accessible file) 2. (temporary directory)/file.name, 3. /target_project_space_dir/filename (in case the user gave a path relative to the target project space). Only if no file of the same name is found, the file is retrieved from the fileserver.
```
Can you introduce some line breaks here?

thawn replied:
done

haesleinhuepf reviewed (changes requested):
Hi Till @thawn ,

that's super cool! I'm tempted to merge and try it immediately. There are just some minor things I would like to ask for improving code-quality. Otherwise it looks amazing!

Thanks!
Robert

thawn commented:
thanks a lot for the code review. I will implement the changes you suggested when I am back from vacation (today I will focus on resubmitting a paper).
Cheers,
Till

thawn committed:
added requirements

thawn committed:
applied changes suggested by @haesleinhuepf

thawn committed:
fixed rsync path

thawn commented:
@haesleinhuepf I now implemented the changes you suggested.
I left two comments open for you to check if the way I implemented the changes is o.k. with you.

thawn committed:
added generic save_to_project function

save_to_project takes the save function as argument, so it can be used
to save any kind of data like numpy arrays or pandas tables etc.

thawn committed:
started unit tests

thawn committed:
added convenience functions

... for loading and saving data other than images
save and load modules are now imported as they are needed, not before

thawn committed:
test_ensure_project_dir_exists

thawn committed:
added _ensure_project_dir_exists

thawn committed:
explicitly cleanup tmp

thawn committed:
improved code readability

thawn committed:
expose ConfirmationRequiredException

thawn committed:
more consistent output of remove_file

thawn committed:
added pandas load and save methods

use pandas instead of numpy for csv_load and csv_save

thawn committed:
added pandas excel and to_hdf

thawn committed:
added tests

thawn committed:
explicitly turned off quiet mode for waitfor

thawn committed:
added quiet parameter to ProjectFileTransfer

haesleinhuepf commented on tests/test.py:
```diff
@@ -0,0 +1,319 @@
+except ModuleNotFoundError:
+    import sys
+    sys.path.append(str(my_path.parent.parent))
+    import biapol_taurus
```
That's interesting! Is it a workaround? And if yes, for what? Please put a comment or remove it :-)

haesleinhuepf commented on tests/test.py:
```diff
@@ -0,0 +1,319 @@
+            self.assertTrue(np.array_equal(self.testdata, loaded))
+
+
+if __name__ == '__main__':
+    unittest.main()
```
Is this necessary? Why not calling tests using `pytest`?

haesleinhuepf commented on biapol_taurus/_project_file_transfer.py:
```diff
@@ -1,186 +1,652 @@
+        full_path = self.get_file(filename)
+        return pd_read_hdf(str(full_path), *args, **kw)
+
+    def pandas_to_hdf(self, filename, data, *args, target: str = 'project', **kw):
```
I would like to avoid hdf files, because they don't open in any other software. We could keep that function if it has a `_`  at the beginning of its name

haesleinhuepf reviewed (approved):
Just some minor comments

thawn committed:
added optional dependencies

thawn committed:
use existing tempdir

thawn committed:
fixed issue with missing temp dir

thawn committed:
fixes issue #3

thawn committed:
make sure that the temporary directory exists

thawn committed:
added sync_from_fileserver and sync_to_fileserver

closes issue #5

thawn committed:
re-ordered methods

thawn committed:
make function signatures of save and load functions identical to the original functions

thawn committed:
mark list_fileserver_files as private

hack to address issue #9

thawn committed:
ProjectFileTransfer now always uses /scratch

use either fileserver or project space as source.
no more syncing between project space and fileserver

thawn committed:
marked save_file and load_file as internal functions

thawn committed:
added get_file and put_file

thawn committed:
re-ordered public and private methods

thawn committed:
removed obsolete imports

thawn committed:
use get_test_cluster

thawn committed:
fixed regex

thawn committed:
proper cleanup

thawn committed:
switch from deprecated xls to xlsx format

thawn committed:
clearer function names

thawn committed:
improve sync output format

thawn committed:
improved docstring

thawn committed:
explicitly declare temporary_directory_path

thawn committed:
do not sync the temporary directory

thawn committed:
added cleanup method

thawn committed:
also check the cache in _load_file

avoid unnecessary file transfers from the fileserver

thawn committed:
added test that ensures synced files are not transferred again

thawn committed:
fixed relative path in cache

thawn committed:
added jupyter notebook for testing

thawn committed:
use separate workspace for tmp

fixes issue #11

thawn committed:
adapted tests to new tmp workspace

thawn committed:
re-ran test notebook on cluster

thawn committed:
added documentation to test notebook

thawn committed:
Update setup.py
